plugins/parade-lspcon/fu-parade-lspcon-device.c

```diff
@@ -0,0 +1,829 @@
+			     device_name);
+		return FALSE;
+	}
+
+	/* should know which aux device over which we read DPCD version */
+	if (self->aux_device_name == NULL) {
+		g_set_error_literal (error,
+				     FWUPD_ERROR,
+				     FWUPD_ERROR_NOT_SUPPORTED,
+				     "ParadeLspconAuxDeviceName must be specified");
+		return FALSE;
+	}
+
+	/* FuI2cDevice->probe */
+	return FU_DEVICE_CLASS (fu_parade_lspcon_device_parent_class)->probe (device, error);
+}
+
+static gboolean
+fu_parade_lspcon_ensure_i2c_address (FuParadeLspconDevice *self,
+				     guint8 address,
+				     GError **error)
+{
+	if (!fu_udev_device_ioctl (FU_UDEV_DEVICE (self),
+				   I2C_SLAVE, (guint8 *) (guintptr) address, NULL,
+				   error)) {
+		g_prefix_error (error, "failed to set I2C slave address: ");
+		return FALSE;
+	}
+	return TRUE;
+}
+
+static gboolean
+fu_parade_lspcon_device_open (FuDevice *device, GError **error)
+{
+	if (!FU_DEVICE_CLASS (fu_parade_lspcon_device_parent_class)->open (device, error))
+		return FALSE;
+
+	/* general assumption is that page 2 is selected: code that uses another address
+	 * should use an address guard to ensure it gets reset */
+	return fu_parade_lspcon_ensure_i2c_address (FU_PARADE_LSPCON_DEVICE (device),
+						    I2C_ADDR_PAGE2,
+						    error);
+}
+
+/**
+ * creates a scope in which the device's target I2C address is something
+ * other than page 2, and resets it to page 2 when the scope is left.
+ */
+typedef struct {
+	FuParadeLspconDevice *device;
+} FuParadeLspconI2cAddressGuard;
+
+static FuParadeLspconI2cAddressGuard *
+fu_parade_lspcon_i2c_address_guard_new (FuParadeLspconDevice *self, guint8 new_address, GError **error)
+{
+	FuParadeLspconI2cAddressGuard *out;
+
+	if (!fu_parade_lspcon_ensure_i2c_address (self, new_address, error))
+		return NULL;
+	out = g_new0 (FuParadeLspconI2cAddressGuard, 1);
+	out->device = self;
+	return out;
+}
+
+static void
+fu_parade_lspcon_i2c_address_guard_free (FuParadeLspconI2cAddressGuard *guard)
+{
+	g_autoptr(GError) error_local = NULL;
+	if (!fu_parade_lspcon_ensure_i2c_address (guard->device,
+						  I2C_ADDR_PAGE2,
+						  &error_local)) {
+		g_warning ("failed to set page2 back: %s", error_local->message);
+	}
+	g_free (guard);
+}
+
+G_DEFINE_AUTOPTR_CLEANUP_FUNC(FuParadeLspconI2cAddressGuard,
+			      fu_parade_lspcon_i2c_address_guard_free);
+
+static gboolean
+fu_parade_lspcon_write_register (FuParadeLspconDevice *self,
+				 guint8 register_addr,
+				 guint8 value,
+				 GError **error)
+{
+	guint8 transaction[] = { register_addr, value };
+	return fu_i2c_device_write_full (FU_I2C_DEVICE (self), transaction,
+					 sizeof(transaction), error);
+}
+
+static gboolean
+fu_parade_lspcon_read_register (FuParadeLspconDevice *self,
+				guint8 register_addr,
+				guint8 *value,
+				GError **error)
+{
+	FuI2cDevice *i2c_device = FU_I2C_DEVICE (self);
+	if (!fu_i2c_device_write (i2c_device, register_addr, error))
+		return FALSE;
+	return fu_i2c_device_read (i2c_device, value, error);
```

Comment:
is this a standard i2c thing? i.e. should we rename `fu_parade_lspcon_read_register()` to `fu_i2c_device_read_register()`?

Reply:
This is a common pattern, but not standardized anywhere. It might be reasonable to provide, though a name like `fu_i2c_device_write_read` might be better since it doesn't imply any particular semantics.

Reply:
Okay, I'll leave it until there's another user that does the same thing; rule of three and all that.